text/0000-ancillary-package-module-standardization.md

```diff
@@ -0,0 +1,234 @@
+is composed of many different small packages.  Due to fragmentation in
+the JavaScript ecosystem, these packages use different build pipelines,
+module formats, output formats, and naming conventions.  This
+fragmentation creates significant overhead for contributors, as they
+must familiarize themselves with the quirks of each package's build
+system and module format. Typically, this information is severely
+underdocumented if it's documented at all.
+
+Additionally, this complexity leaks out to anyone who wishes to consume
+these packages. Rather than adhering to current conventions and best
+practices for distributing code on npm, users must learn the vagaries of
+how to consume each package; whether that's by consuming a global,
+consuming anonymous AMD, named AMD, or something else. Ember's build
+system accrues needless complexity because it has to handle all of these
+different formats distributed in different ways.
+
+Lastly, due to efforts like FastBoot, users expect to be able to run
+their JavaScript code both in the browser and in Node.js.  We need a
+system that makes our packages portable, so that they can be tested and
+run in any environment.
+
+# Detailed Design
+
+This RFC proposes that all packages under the Ember.js umbrella be
+incrementally migrated to standard JavaScript (ES6) modules.  This
+format is trivially transpilable into whatever ultimate module system
+the end user desires, whether that's {named, anonymous} AMD, CommonJS,
+or
+no module system at all using something like Rollup's bundle builds.
+
+JavaScript modules have the benefit of being statically analyzable, so
+tools can determine the dependency graph at build time.  By normalizing
+Ember's dependencies into JavaScript modules, we make it much easier to
+eliminate code that is not actually needed in final builds.
+
+To facilitate a backwards-compatible migration, we will standardize on a
+directory layout for projects, as well as a directory layout for the
+final built output. All packages will be distributed via npm.
+
+To ensure standardized behavior across repositories, a reusable npm
+package `broccoli-module-alchemist` will be created to encapsulate the
+decisions of this RFC.
+
+The source code for broccoli-module-alchemist is available at
+<https://github.com/monegraph/broccoli-module-alchemist>.
+
+## Supported Authoring Formats
+
+Authors may write package source code in TypeScript, ES2015, or
+a mix of both. Because transpilation is done using the TypeScript
+compiler, which supports both, projects can incrementally migrate
+towards TypeScript, or use it only in portions where it makes sense.
+
+Whether using TypeScript or JavaScript, authors should always use
+standard ES2015 module syntax for both importing and exporting modules.
+
+## Target Formats
+
+The original source code is compiled from TypeScript or ES2015+ to ES5
+to ensure maximum compatibility across browser and Node.js versions.
```

Comment:
> @martndemus I was just discussing this with @dgeb. My feeling is that consumers who wanted latest JavaScript would just consume src/, but that would mean all consumers have to understand TypeScript, for example.

I'm not sure this is sufficient, as the `src/` will not just be es2015 CODE but could be anything. TS/Flow/Dart and can be decorated with dev instrumentation and asserts.

I suspect what consumers of ES2015 will want, is actually (most likely) also the result of some build step. Which leads me to believe, we should consider the *baseline target* being configurable, and by default we can guarantee ES5 and ES2015, but over time add ES2016 and so an so forth.

So What if we lifted the constraint slightly, forcing all our micro-libs to support something like:

```js
function buildMyMicroLib(options) {
  options.target // => 2015 || ES5 || ...

  return produceForTarget(tree, target);
}
```

That way:
* for our build tools, we can request the baseline target and the tool will provide.
* For others not using our build tools, we can pre-compute the various outputs they need (maybe one per baseline)

--

I guess my above thoughts are inspire by things like typings, asserts, instrumentation and likely other build time concerns source, which leave our source unconsummable without a build step.